packages/lesswrong/lib/search/NativeSearchClient.ts

```diff
@@ -0,0 +1,80 @@
+import type { Client, Index } from "algoliasearch/lite";
+import type { MultiResponse, QueryParameters, SearchForFacetValues } from "algoliasearch";
+import LRU from "lru-cache";
+
+export type SearchQuery = {
+  indexName: string;
+  query: string;
+  params: QueryParameters;
+}
+
+class NativeSearchClient implements Client {
+  private headers: Record<string, string> = {};
+  private cache = new LRU<string, Promise<MultiResponse<unknown>>>();
+
+  initIndex(_indexName: string): Index {
+    throw new Error("initIndex not supported by NativeSearchClient");
+  }
+
+  search<T=unknown>(
+    queries: SearchQuery[],
+    cb: (err: Error|null, res: MultiResponse<T>|null) => void,
+  ): void;
+  search<T=unknown>(queries: SearchQuery[]): Promise<MultiResponse<T>>;
+  search<T=unknown>(
+    queries: SearchQuery[],
+    cb?: (err: Error|null, res: MultiResponse<T>|null) => void,
+  ): Promise<MultiResponse<T>>|void {
+    const body = JSON.stringify(queries);
+    const cached = this.cache.get(body);
+    if (cached) {
+      return Promise.resolve(cached) as Promise<MultiResponse<T>>;
+    }
+    const promise = new Promise<MultiResponse<T>>((resolve, reject) => {
+      fetch("/api/search", {
+        method: "POST",
+        headers: {
+          ...this.headers,
+          "Content-Type": "application/json",
+        },
+        body,
+      }).then((response) => {
+        response.json().then((results) => {
+          resolve({results});
+        }).catch(reject);
+      }).catch(reject);
+    });
+    this.cache.set(body, promise);
```

Comment:
`JSON.stringify` isn't deterministic, so if we're using a stringified version of the queries as a cache key we should maybe using something like (json-stringify-determinstic)[https://www.npmjs.com/package/json-stringify-deterministic].

I'm actually quite unsure about having a cache here at all, unless there's an extremely fat tail of queries such that we get a decent cache hit rate even with a short TTL (speaking of which - what's the default TTL in `lru-cache`?).  Having a long TTL could be problematic for search results, since we're not doing any manual invalidation and it doesn't seem great to be serving old versions of posts or comments that might have been edited after they were created.

Reply:
Ok, if it's client-side with a short TTL that seems fine.